File: desktop-tauri/src-tauri/src/xlsx_rpc.rs
//! `xlsx_rpc` — the Tauri-mode half of the sheets xlsx-RPC channel (the
//! browser-mode half is `tools/xlsx-sidecar-server.mjs`; the webview client is
//! `src/bridge/xlsx-rpc.ts`).
//!
//! One long-running xlsx-sidecar child process, newline-delimited JSON on
//! stdio, requestId matching, 30s timeouts (180s for archive commands) —
//! mirroring the upstream Electron client
//! `desktop/apps/sheets/src/main/xlsx-sidecar-client.ts`. Commands starting
//! with `host.` are answered locally: they are the fs touchpoints the
//! in-webview gateway save pipeline needs (temp dirs, plan content files,
//! staging URL-ish paths so the sidecar sees real files).

use std::collections::HashMap;
use std::path::Path;
use std::process::Stdio;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use base64::Engine;
use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, ChildStdin, Command};
use tokio::sync::{oneshot, Mutex};

const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);
/// Archive commands stream whole workbooks; large files need more headroom.
const ARCHIVE_TIMEOUT: Duration = Duration::from_secs(180);
const ARCHIVE_COMMANDS: [&str; 6] = [
    "convert_workbook",
    "archive_manifest",
    "read_entries",
    "scan_entries",
    "save_archive",
    "recalc_cells",
];

type PendingMap = Arc<Mutex<HashMap<String, oneshot::Sender<Result<Value, String>>>>>;

struct SidecarChild {
    stdin: Mutex<ChildStdin>,
    pending: PendingMap,
}

/// Process-wide sidecar state (one workbook engine per app run).
#[derive(Default)]
pub struct XlsxRpcState {
    child: Mutex<Option<Arc<SidecarChild>>>,
}

fn sidecar_path() -> Result<String, String> {
    if let Ok(path) = std::env::var("XLSX_SIDECAR_PATH") {
        if !path.is_empty() {
            return Ok(path);
        }
    }
    let relative = Path::new("native")
        .join("xlsx-engine")
        .join("target")
        .join("release")
        .join("xlsx-sidecar");
    // dev: cwd is src-tauri, so the binary lives one level up; bundled
    // layouts are a bundling decision (TODO M3: tauri sidecar resource).
    let mut candidates = vec![relative.clone()];
    if let Ok(cwd) = std::env::current_dir() {
        candidates.push(cwd.join("..").join(&relative));
        if let Ok(exe) = std::env::current_exe() {
            if let Some(dir) = exe.parent() {
                candidates.push(dir.join(&relative));
                candidates.push(dir.join("../../..").join(&relative));
            }
        }
    }
    for candidate in candidates {
        if candidate.is_file() {
            return Ok(candidate.to_string_lossy().into_owned());
        }
    }
    Err("xlsx-sidecar binary not found (set XLSX_SIDECAR_PATH or build native/xlsx-engine)".into())
}

fn fail_all(pending: &PendingMap, message: String) {
    let pending = pending.clone();
    tauri::async_runtime::spawn(async move {
        let mut map = pending.lock().await;
        for (_, sender) in map.drain() {
            let _ = sender.send(Err(message.clone()));
        }
    });
}

async fn ensure_child(state: &XlsxRpcState) -> Result<Arc<SidecarChild>, String> {
    let mut guard = state.child.lock().await;
    if let Some(child) = guard.as_ref() {
        return Ok(child.clone());
    }
    let binary = sidecar_path()?;
    let mut process: Child = Command::new(binary)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .map_err(|e| format!("failed to spawn xlsx-sidecar: {e}"))?;
    let stdin = process.stdin.take().ok_or("xlsx-sidecar stdin unavailable")?;
    let stdout = process.stdout.take().ok_or("xlsx-sidecar stdout unavailable")?;
    let pending: PendingMap = Arc::new(Mutex::new(HashMap::new()));
    let reader_pending = pending.clone();
    tauri::async_runtime::spawn(async move {
        let mut lines = BufReader::new(stdout).lines();
        loop {
            match lines.next_line().await {
                Ok(Some(line)) => {
                    let parsed: Result<Value, _> = serde_json::from_str(&line);
                    match parsed {
                        Ok(response) => {
                            let request_id = response
                                .get("requestId")
                                .and_then(Value::as_str)
                                .unwrap_or("")
                                .to_string();
                            let sender = reader_pending.lock().await.remove(&request_id);
                            if let Some(sender) = sender {
                                if response.get("ok").and_then(Value::as_bool) == Some(true) {
                                    let _ =
                                        sender.send(Ok(response.get("result").cloned().unwrap_or(Value::Null)));
                                } else {
                                    let message = response
                                        .get("error")
                                        .and_then(|e| e.get("message"))
                                        .and_then(Value::as_str)
                                        .unwrap_or("XLSX sidecar request failed.")
                                        .to_string();
                                    let _ = sender.send(Err(message));
                                }
                            }
                        }
                        Err(_) => {
                            fail_all(
                                &reader_pending,
                                "XLSX sidecar returned invalid JSON.".to_string(),
                            );
                        }
                    }
                }
                Ok(None) => {
                    fail_all(&reader_pending, "XLSX sidecar exited.".to_string());
                    break;
                }
                Err(error) => {
                    fail_all(&reader_pending, format!("XLSX sidecar read failed: {error}"));
                    break;
                }
            }
        }
    });
    let child = Arc::new(SidecarChild {
        stdin: Mutex::new(stdin),
        pending,
    });
    *guard = Some(child.clone());
    Ok(child)
}

/// After a child death the next request respawns; called on write/timeout failure.
async fn drop_child(state: &XlsxRpcState, child: &Arc<SidecarChild>) {
    let mut guard = state.child.lock().await;
    if guard.as_ref().is_some_and(|c| Arc::ptr_eq(c, child)) {
        *guard = None;
    }
}

/// FNV-1a hex — stable staging key for a logical path (no extra crates).
fn fnv_hex(input: &str) -> String {
    let mut hash: u64 = 0xcbf29ce484222325;
    for byte in input.as_bytes() {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(0x100000001b3);
    }
    format!("{hash:016x}")
}

static TEMP_SEQ: AtomicU64 = AtomicU64::new(0);

fn unique_temp_dir(prefix: &str) -> std::path::PathBuf {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let seq = TEMP_SEQ.fetch_add(1, Ordering::Relaxed);
    std::env::temp_dir().join(format!("{prefix}{}-{}-{seq}", std::process::id(), nanos))
}

fn sanitize_name(name: &str) -> String {
    let cleaned: String = name
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() || matches!(c, '.' | '_' | '-') { c } else { '_' })
        .collect();
    if cleaned.is_empty() {
        "workbook.xlsx".to_string()
    } else {
        cleaned
    }
}

async fn handle_host(command: &str, request: &Value) -> Result<Value, String> {
    let str_arg = |key: &str| -> Result<String, String> {
        request
            .get(key)
            .and_then(Value::as_str)
            .map(str::to_string)
            .ok_or_else(|| format!("{command}: missing string argument '{key}'"))
    };
    match command {
        // Real on-disk paths pass through; anything else (vite URL-ish paths,
        // overlay keys) is staged under a temp dir keyed by the logical path.
        "host.stage" => {
            let logical = str_arg("path")?;
            let path = Path::new(&logical);
            if path.is_absolute() && path.is_file() {
                return Ok(serde_json::json!({ "path": logical }));
            }
            let dir = std::env::temp_dir()
                .join("panoffice-xlsx-stage")
                .join(&fnv_hex(&logical)[..16]);
            tokio::fs::create_dir_all(&dir).await.map_err(|e| e.to_string())?;
            let name = path
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default();
            let staged = dir.join(sanitize_name(&name));
            let bytes = base64::engine::general_purpose::STANDARD
                .decode(request.get("base64").and_then(Value::as_str).unwrap_or(""))
                .map_err(|e| format!("host.stage: bad base64: {e}"))?;
            tokio::fs::write(&staged, bytes).await.map_err(|e| e.to_string())?;
            Ok(serde_json::json!({ "path": staged.to_string_lossy() }))
        }
        "host.mkdtemp" => {
            let prefix = request
                .get("prefix")
                .and_then(Value::as_str)
                .unwrap_or("panoffice-xlsx-");
            let dir = unique_temp_dir(prefix);
            tokio::fs::create_dir_all(&dir).await.map_err(|e| e.to_string())?;
            Ok(serde_json::json!({ "path": dir.to_string_lossy() }))
        }
        "host.mkdir" => {
            tokio::fs::create_dir_all(str_arg("path")?)
                .await
                .map_err(|e| e.to_string())?;
            Ok(serde_json::json!({}))
        }
        "host.read_text" => {
            let text = tokio::fs::read_to_string(str_arg("path")?)
                .await
                .map_err(|e| e.to_string())?;
            Ok(serde_json::json!({ "text": text }))
        }
        "host.read_file" => {
            let bytes = tokio::fs::read(str_arg("path")?)
                .await
                .map_err(|e| e.to_string())?;
            Ok(serde_json::json!({
                "base64": base64::engine::general_purpose::STANDARD.encode(bytes)
            }))
        }
        "host.write_file" => {
            let path = str_arg("path")?;
            if let Some(text) = request.get("text").and_then(Value::as_str) {
                tokio::fs::write(&path, text).await.map_err(|e| e.to_string())?;
            } else {
                let bytes = base64::engine::general_purpose::STANDARD
                    .decode(request.get("base64").and_then(Value::as_str).unwrap_or(""))
                    .map_err(|e| format!("host.write_file: bad base64: {e}"))?;
                tokio::fs::write(&path, bytes).await.map_err(|e| e.to_string())?;
            }
            Ok(serde_json::json!({}))
        }
        "host.remove" => {
            let path = str_arg("path")?;
            let recursive = request.get("recursive").and_then(Value::as_bool) == Some(true);
            let result = if recursive {
                tokio::fs::remove_dir_all(&path).await
            } else {
                tokio::fs::remove_file(&path).await
            };
            // force semantics: a missing path is not an error
            match result {
                Ok(()) => Ok(serde_json::json!({})),
                Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(serde_json::json!({})),
                Err(e) => Err(e.to_string()),
            }
        }
        "host.rename" => {
            tokio::fs::rename(str_arg("from")?, str_arg("to")?)
                .await
                .map_err(|e| e.to_string())?;
            Ok(serde_json::json!({}))
        }
        _ => Err(format!("Unknown host command: {command}")),
    }
}

#[tauri::command]
pub async fn xlsx_rpc(
    state: tauri::State<'_, XlsxRpcState>,
    request: Value,
) -> Result<Value, String> {
    let command = request
        .get("command")
        .and_then(Value::as_str)
        .ok_or("xlsx_rpc: missing command")?
        .to_string();
    let request_id = request
        .get("requestId")
        .and_then(Value::as_str)
        .ok_or("xlsx_rpc: missing requestId")?
        .to_string();
    if let Some(host_command) = command.strip_prefix("host.") {
        let _ = host_command;
        return handle_host(&command, &request).await;
    }

    let child = ensure_child(&state).await?;
    let (tx, rx) = oneshot::channel();
    child.pending.lock().await.insert(request_id.clone(), tx);
    let line = format!("{}\n", serde_json::to_string(&request).map_err(|e| e.to_string())?);
    {
        let mut stdin = child.stdin.lock().await;
        if let Err(error) = stdin.write_all(line.as_bytes()).await {
            child.pending.lock().await.remove(&request_id);
            drop_child(&state, &child).await;
            return Err(format!("XLSX sidecar write failed: {error}"));
        }
    }
    let limit = if ARCHIVE_COMMANDS.contains(&command.as_str()) {
        ARCHIVE_TIMEOUT
    } else {
        REQUEST_TIMEOUT
    };
    match tokio::time::timeout(limit, rx).await {
        Ok(Ok(result)) => result,
        Ok(Err(_dropped)) => {
            drop_child(&state, &child).await;
            Err("XLSX sidecar exited.".to_string())
        }
        Err(_elapsed) => {
            child.pending.lock().await.remove(&request_id);
            Err("XLSX sidecar request timed out.".to_string())
        }
    }
}

File: desktop-tauri/src/bridge/xlsx-rpc.ts
/**
 * xlsx-RPC client — the webview half of the sheets sidecar channel.
 *
 * One request/response envelope for two transports:
 *  - Tauri: `invoke('xlsx_rpc', { request })` — the Rust command owns the
 *    sidecar child (spawn/reuse, newline-JSON framing, requestId matching,
 *    30s/180s timeouts) and answers `host.*` commands itself,
 *  - browser (vite dev / headless e2e): HTTP POST to the dev server
 *    (tools/xlsx-sidecar-server.mjs), which does the same in Node.
 *
 * `host.*` commands are the fs touchpoints the gateway save pipeline needs
 * (temp dirs, plan content files, staging logical paths so the sidecar sees
 * real files); everything else is forwarded to the sidecar verbatim. The
 * semantics mirror desktop/apps/sheets/src/main/xlsx-sidecar-client.ts.
 */
import { invoke } from '@tauri-apps/api/core'

import type { ArchiveClient } from '../apps/sheets/gateway/xlsx-package-io'
import type { HostIo } from '../apps/sheets/gateway/host-io'
import { base64ToBytes, bytesToBase64 } from '../apps/sheets/gateway/sha256'
import { isTauri } from './platform'

const PROTOCOL_VERSION = 1
const REQUEST_TIMEOUT_MS = 30_000
/** Archive commands stream whole workbooks; large files need more headroom. */
const ARCHIVE_TIMEOUT_MS = 180_000
/** Webview-side backstop on top of the host's own timeout (host errors win). */
const TIMEOUT_MARGIN_MS = 10_000

const ARCHIVE_COMMANDS = new Set([
  'convert_workbook',
  'archive_manifest',
  'read_entries',
  'scan_entries',
  'save_archive',
  'recalc_cells',
  // answered by the WOPI host itself: server-side staging of a store file
  'host.stage_store_file',
])

/** XLSX-service base URL; production uses the WOPI host's same-origin proxy. */
const SERVER_URL =
  (import.meta.env.VITE_XLSX_SIDECAR_URL as string | undefined) ?? 'http://127.0.0.1:8791'

const errMsg = (err: unknown) => (err instanceof Error ? err.message : String(err))

export interface XlsxRpcResponse {
  readonly ok: boolean
  readonly result?: unknown
  readonly error?: { readonly code?: string; readonly message?: string }
}

/**
 * Send one command down the channel; resolves with the sidecar/host `result`
 * or rejects with its error message, exactly like the upstream Node client.
 */
export async function xlsxRpc(command: Readonly<Record<string, unknown>>): Promise<unknown> {
  const kind = typeof command.command === 'string' ? command.command : ''
  const timeoutMs = (ARCHIVE_COMMANDS.has(kind) ? ARCHIVE_TIMEOUT_MS : REQUEST_TIMEOUT_MS) +
    TIMEOUT_MARGIN_MS
  const envelope = {
    version: PROTOCOL_VERSION,
    requestId: crypto.randomUUID(),
    ...command,
  }
  const response = await withTimeout(
    isTauri() ? tauriRoundTrip(envelope) : httpRoundTrip(envelope),
    timeoutMs,
  )
  if (response.ok) return response.result
  throw new Error(response.error?.message ?? 'XLSX sidecar request failed.')
}

async function tauriRoundTrip(envelope: Readonly<Record<string, unknown>>): Promise<XlsxRpcResponse> {
  // Rust returns the response payload (or throws the sidecar's error message);
  // wrapping it keeps one response shape for both transports.
  try {
    const result = await invoke<unknown>('xlsx_rpc', { request: envelope })
    return { ok: true, result }
  } catch (err) {
    return { ok: false, error: { message: errMsg(err) } }
  }
}

async function httpRoundTrip(envelope: Readonly<Record<string, unknown>>): Promise<XlsxRpcResponse> {
  let res: Response
  try {
    res = await fetch(`${SERVER_URL}/rpc`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(envelope),
    })
  } catch (err) {
    throw new Error(
      `XLSX service unreachable at ${SERVER_URL}: ${errMsg(err)}`,
    )
  }
  const parsed = (await res.json().catch(() => null)) as {
    ok?: boolean
    result?: unknown
    error?: { message?: string }
  } | null
  if (!res.ok) {
    throw new Error(parsed?.error?.message ?? `XLSX service answered HTTP ${res.status}`)
  }
  if (parsed?.ok === true) return { ok: true, result: parsed.result }
  return { ok: false, error: { message: parsed?.error?.message ?? 'XLSX request failed.' } }
}

function withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
  let timer: ReturnType<typeof setTimeout> | undefined
  const timeout = new Promise<never>((_resolve, reject) => {
    timer = setTimeout(() => reject(new Error('XLSX sidecar request timed out.')), timeoutMs)
  })
  return Promise.race([promise, timeout]).finally(() => clearTimeout(timer))
}

// ---- typed wrappers (mirror XlsxSidecarClient's method surface) ----

export const xlsxClient = {
  open: (path: string): Promise<unknown> => xlsxRpc({ command: 'open', path }),
  readRange: (input: {
    sessionId: string
    sheetId: string
    range: { startRow: number; endRow: number; startColumn: number; endColumn: number }
  }): Promise<unknown> => xlsxRpc({ command: 'read_range', ...input }),
  readFormulaCells: (input: { sessionId: string; sheetId: string }): Promise<unknown> =>
    xlsxRpc({ command: 'read_formula_cells', ...input }),
  readMedia: (input: { sessionId: string; visualId: string }): Promise<unknown> =>
    xlsxRpc({ command: 'read_media', ...input }),
  recalcCells: (input: {
    path: string
    edits: readonly { sheet: string; row: number; column: number; input: string }[]
    reads: readonly {
      sheet: string
      range: { startRow: number; endRow: number; startColumn: number; endColumn: number }
    }[]
  }): Promise<unknown> => xlsxRpc({ command: 'recalc_cells', ...input }),
  convertWorkbook: (input: { path: string; targetPath: string }): Promise<unknown> =>
    xlsxRpc({ command: 'convert_workbook', ...input }),
  close: (sessionId: string): Promise<unknown> => xlsxRpc({ command: 'close', sessionId }),
}

/** ArchiveClient over the RPC channel — the gateway's save path consumes this. */
export function createRpcArchiveClient(): ArchiveClient {
  return {
    archiveManifest: (path) => xlsxRpc({ command: 'archive_manifest', path }),
    readEntries: (input) => xlsxRpc({ command: 'read_entries', ...input }),
    scanEntries: (input) => xlsxRpc({ command: 'scan_entries', ...input }),
    saveArchive: (input) => xlsxRpc({ command: 'save_archive', ...input }),
  }
}

async function hostCall(command: string, args: Record<string, unknown>): Promise<unknown> {
  return xlsxRpc({ command, ...args })
}

/** HostIo over the RPC channel — temp dirs and plan content files on the host. */
export function createRpcHostIo(): HostIo {
  return {
    mkdtemp: async (prefix) => {
      const result = (await hostCall('host.mkdtemp', { prefix })) as { path: string }
      return result.path
    },
    mkdir: async (path) => {
      await hostCall('host.mkdir', { path })
    },
    readText: async (path) => {
      const result = (await hostCall('host.read_text', { path })) as { text: string }
      return result.text
    },
    writeFile: async (path, content) => {
      if (typeof content === 'string') await hostCall('host.write_file', { path, text: content })
      else await hostCall('host.write_file', { path, base64: bytesToBase64(content) })
    },
    remove: async (path, opts) => {
      await hostCall('host.remove', { path, recursive: opts?.recursive ?? false })
    },
    rename: async (from, to) => {
      await hostCall('host.rename', { from, to })
    },
  }
}

/**
 * Ensure the logical path the webview reads/writes (overlay key or real path)
 * exists as a real file the sidecar can open; returns the host-side path.
 * Tauri passes real paths through untouched; the dev server stages the bytes
 * under a temp directory keyed by the logical path.
 */
export async function stageForSidecar(logicalPath: string, bytes: Uint8Array): Promise<string> {
  const result = (await hostCall('host.stage', {
    path: logicalPath,
    base64: bytesToBase64(bytes),
  })) as { path: string }
  return result.path
}

/** Read a host-side file back into the webview (saved workbook bytes). */
export async function readHostFile(path: string): Promise<Uint8Array> {
  const result = (await hostCall('host.read_file', { path })) as { base64: string }
  return base64ToBytes(result.base64)
}
